File: Data Structures & Algorithms/add-two-numbers/submission-5.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} l1
     * @param {ListNode} l2
     * @return {ListNode}
     */
    addTwoNumbers(l1, l2) {
        const dummy = new ListNode();
        this.addTwoNumbersRecursive(l1, l2, 0, dummy);

        return dummy.next;

        // let curr = dummy;

        // let carry = 0;

        // while (l1 != null || l2 != null || carry > 0) {
        //     const v1 = l1 === null ? 0 : l1.val;
        //     const v2 = l2 === null ? 0 : l2.val;
        //     let val = v1 + v2 + carry;
        //     carry = Math.trunc(val / 10);
        //     val = val % 10;

        //     curr.next = new ListNode(val);
        //     curr = curr.next;

        //     l1 = l1 === null ? null : l1.next;
        //     l2 = l2 === null ? null : l2.next;
        // }

        // return dummy.next;
    }

    addTwoNumbersRecursive(l1, l2, carry, curr) {
        if (l1 === null && l2 === null) {
            if (carry > 0) {
                const newNode = new ListNode(1);
                curr.next = newNode;

                return newNode;
            } else {
                return null;
            }
        }

        const v1 = l1 === null ? 0 : l1.val;
        const v2 = l2 === null ? 0 : l2.val;
        let val = v1 + v2 + carry;
        carry = Math.trunc(val / 10);
        val = val % 10;

        const newNode = new ListNode(val);
        curr.next = newNode

        newNode.next = this.addTwoNumbersRecursive(l1 === null ? null : l1.next, l2 === null ? null : l2.next, carry, newNode);

        return newNode;
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-5.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        // edge case where head is null or single
        if (head == null || head.next == null) {
            return head;
        }

        // get size of linked list
        ListNode curr = head;
        int size = 0;
        while (curr != null) {
            curr = curr.next;
            size++;
        }

        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode prev = dummy;
        curr = head;

        System.out.println("size " + size + " k " + k);

        for (int i = 0; i < size / k; i++) {
            ListNode test = dummy.next;
            while (test != null) {
                System.out.println("l: " + test.val);
                test = test.next;
            }


            System.out.println("prev = " + prev.val + " curr = " + curr.val);
            ListNode oldTail = curr;
            for (int j = 0; j < k - 1; j++) {
                System.out.println("oT " + oldTail.val);
                oldTail = oldTail.next;
            }
            ListNode nextNode = oldTail.next;
            System.out.println("prev = " + prev.val + " curr = " + curr.val + " oldTail = " + oldTail.val);

            // cut off connections with prev and nextNode
            // prev.next = curr right now
            // oldTail.next = nextNode right now
            prev.next = null;
            oldTail.next = null;

            // reverse the k-size list
            ListNode[] heads = reverseList(curr);
            
            // re-set prev.next to be newHead [index 1]
            prev.next = heads[1];

            // re-set new tail to nextNode
            heads[0].next = nextNode;

            curr = prev.next;

            // now move everything over by k steps
            for (int j = 0; j < k; j++) {
                prev = prev.next;
                curr = curr.next;
            }
        }

        return dummy.next;
    }

    // returns the [oldHead, newHead] where oldHead is now the tail, 
    // and newHead is the newHead of the reversed list
    private ListNode[] reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;

        while (curr != null) {
            ListNode next = curr.next;

            curr.next = prev;
            prev = curr;
            curr = next;
        }

        return new ListNode[] {head, prev};
    }
}
